### PR TITLE
[WIP] GPU Rendering

### PR DESCRIPTION
**Pull request by philippotto**
### PR Summary

- Rendering for orthogonal mode is done by moving the necessary buckets to the GPU and rendering these buckets via fragment shaders
- Buckets are written into a data texture whose can be looked up via a lookup-texture. the right slot for a bucket in the lookup texture is found by calculating the difference vector between the bucket's position and between an anchor point, which is derived by the global position.
- Writing new buckets to the data texture is distributed over multiple frames (via 
`requestAnimationFrame`).
- If a bucket is not available to the shader (either because it wasn't moved to the GPU yet or because it wasn't fetched by the server), a fallback bucket of a higher zoomstep is rendered by upscaling it. At the moment, there is only one fallback zoomstep. Ideally, we could use all available zoomsteps as a fallback, but this requires two textures (data + lookup) for each zoomstep. Unfortunately, texture unit count is limited, though.
- The previous CPU rendering was able to use buckets of higher and lower zoomsteps for interpolation. This is why, datasets with only one zoomstep could be looked at while being at a zoom level appropriate for the second zoomstep. With GPU rendering, we cannot use that many buckets for downsampling/upsampling. Instead, we always create another "virtual" zoom step by downsampling the existing buckets to new buckets. This work is done when new buckets are received and is also spread over multiple animation frames.
- The ping strategy was adapted to always load buckets from higher zoomsteps to maximize data available for rendering.
- Three.js was updated to 0.87.0.

### Open issues
- [ ] Support mappings again
- [x] Ensure iPad (and older hardware) compatibility

### Steps to test:
- everything related to rendering (moving, zooming, different settings, such as interpolation)
- data types:
  - "normal" color layers
  - rgb layers
  - segmantation (new segmentation and existing layers)
  - multi layer (e.g., confocal)
- segmentation labeling

### Issues:
- contributes to #2295 

------
- [ ] Ready for review

### CODE REVIEW

File: app/assets/javascripts/libs/UpdatableTexture.js [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import * as THREE from "three";
+
+function UpdatableTexture(
```
Comment (normanrz) [inline app/assets/javascripts/libs/UpdatableTexture.js]: Why no `class` syntax?

File: app/assets/javascripts/oxalis/controller/scene_controller.js [hunk 2]
```diff
@@ -241,12 +242,23 @@ class SceneController {
     const gPos = getPosition(Store.getState().flycam);
     const globalPosVec = new THREE.Vector3(...gPos);
     const planeScale = getPlaneScalingFactor(Store.getState().flycam);
+    let anchorPoint;
```
Comment (normanrz) [inline app/assets/javascripts/oxalis/controller/scene_controller.js]: Perhaps a comment what `anchorPoint` and `fallbackAnchorPoint` are

File: app/assets/javascripts/oxalis/controller/scene_controller.js [hunk 3]
```diff
@@ -241,12 +242,23 @@ class SceneController {
...
+    let fallbackAnchorPoint;
+
+    for (const name of Object.keys(Model.binary)) {
+      const binary = Model.binary[name];
+      const zoomStep = getRequestLogZoomStep(Store.getState());
+      anchorPoint = binary.updateDataTextures(gPos, zoomStep);
+      fallbackAnchorPoint = binary.updateFallbackDataTextures(gPos, zoomStep);
+    }
+
     for (const planeId of OrthoViewValuesWithoutTDView) {
```
Comment (normanrz) [inline app/assets/javascripts/oxalis/controller/scene_controller.js]: `for (const currentPlane of Object.values(this.planes))`

File: app/assets/javascripts/oxalis/model/accessors/flycam_accessor.js [hunk 4]
```diff
@@ -23,18 +23,13 @@ import * as THREE from "three";
...
-const MAX_TEXTURE_OFFSET = 31;
-const MAX_ZOOM_THRESHOLD = 2;
-
-function log2(a: number): number {
-  return Math.log(a) / Math.LN2;
-}
+// Historically, this width decided when which zoom step was.
+// There is no specific reason why this exact size has to be chosen. As long as enough buckets are sent
+// to the GPU (RENDERED_BUCKETS_PER_DIMENSION) this width can be increased or decreased.
+const MAX_RENDING_TARGET_WIDTH = 512 - 32 - 1;
```
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/model/accessors/flycam_accessor.js]: magic numbers

File: app/assets/javascripts/oxalis/model/binary.js [hunk 5]
```diff
@@ -119,15 +112,114 @@ class Binary {
       );
     }
 
-    this.cube.on({
-      newMapping: () => this.forcePlaneRedraw(),
-    });
+    // todo
```
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/model/binary.js]: TODO?

File: app/assets/javascripts/oxalis/model/binary.js [hunk 6]
```diff
@@ -119,15 +112,114 @@ class Binary {
...
+    // this.cube.on({
+    //   newMapping: () => ,
+    // });
+  }
+
+  setupDataTextures(): void {
+    const bytes = this.layer.bitDepth >> 3;
+
+    window.managers = window.managers || [];
```
Comment (hotzenklotz) [inline app/assets/javascripts/oxalis/model/binary.js]: Seems legit  :-P

**Comment by hotzenklotz** [comment]
@philippotto Please also merge the current master so we can test this locally without any problems (slick, postgres etc)

**Comment by philippotto** [comment]
I'm closing this in favor of #2442. Thanks for your feedback, I incorporated most of it in the other PR/branch.

**Pull request closed without merging**